File: project/jimmer-sql/src/main/java/org/babyfish/jimmer/sql/ast/impl/ReplaceExpression.java
package org.babyfish.jimmer.sql.ast.impl;

import org.babyfish.jimmer.sql.ast.Expression;
import org.babyfish.jimmer.sql.ast.impl.render.AbstractSqlBuilder;
import org.jetbrains.annotations.NotNull;

import java.util.Objects;

class ReplaceExpression extends AbstractExpression<String> implements StringExpressionImplementor {

    private Expression<String> raw;
    private final String target;
    private final String replacement;

    ReplaceExpression(Expression<String> raw, String target, String replacement) {
        this.raw = raw;
        this.target = target;
        this.replacement = replacement;
    }

    @Override
    public int precedence() {
        return 0;
    }

    @Override
    public void accept(@NotNull AstVisitor visitor) {
        ((Ast)raw).accept(visitor);
    }

    @Override
    public void renderTo(@NotNull AbstractSqlBuilder<?> builder) {
        builder.sql("replace(");
        ((Ast)raw).renderTo(builder);
        builder.sql(", ");
        builder.rawVariable(target);
        builder.sql(", ");
        builder.rawVariable(replacement);
        builder.sql(")");
    }

    @Override
    protected boolean determineHasVirtualPredicate() {
        return hasVirtualPredicate(raw);
    }

    @Override
    protected Ast onResolveVirtualPredicate(AstContext ctx) {
        raw = ctx.resolveVirtualPredicate(raw);
        return this;
    }

    @Override
    public boolean equals(Object o) {
        if (this == o) return true;
        if (o == null || getClass() != o.getClass()) return false;
        ReplaceExpression that = (ReplaceExpression) o;
        return raw.equals(that.raw) && 
               target.equals(that.target) && 
               replacement.equals(that.replacement);
    }

    @Override
    public int hashCode() {
        return Objects.hash(raw, target, replacement);
    }
} 

File: project/jimmer-sql-kotlin/src/main/kotlin/org/babyfish/jimmer/sql/kt/ast/expression/impl/ReplaceExpression.kt
package org.babyfish.jimmer.sql.kt.ast.expression.impl

import org.babyfish.jimmer.sql.ast.impl.Ast
import org.babyfish.jimmer.sql.ast.impl.AstContext
import org.babyfish.jimmer.sql.ast.impl.AstVisitor
import org.babyfish.jimmer.sql.ast.impl.render.AbstractSqlBuilder
import org.babyfish.jimmer.sql.kt.ast.expression.KExpression
import org.babyfish.jimmer.sql.kt.ast.expression.KNonNullExpression
import org.babyfish.jimmer.sql.kt.ast.expression.KNullableExpression

internal abstract class ReplaceExpression(
    private var raw: KExpression<String>,
    private val target: String,
    private val replacement: String
) : AbstractKExpression<String>() {

    override fun determineHasVirtualPredicate(): Boolean =
        hasVirtualPredicate(raw)

    override fun onResolveVirtualPredicate(ctx: AstContext): Ast? {
        raw = ctx.resolveVirtualPredicate(raw)
        return this
    }

    override fun getType(): Class<String> = String::class.java

    override fun precedence(): Int = 0

    override fun accept(visitor: AstVisitor) {
        (raw as Ast).accept(visitor)
    }

    override fun renderTo(builder: AbstractSqlBuilder<*>) {
        builder.sql("replace(")
        (raw as Ast).renderTo(builder)
        builder.sql(", ")
            .rawVariable(target)
            .sql(", ")
            .rawVariable(replacement)
            .sql(")")
    }

    class NonNull(
        raw: KExpression<String>,
        target: String,
        replacement: String
    ) : ReplaceExpression(raw, target, replacement), KNonNullExpression<String>

    class Nullable(
        raw: KExpression<String>,
        target: String,
        replacement: String
    ) : ReplaceExpression(raw, target, replacement), KNullableExpression<String>
}